packages/flutter/lib/src/widgets/fade_in_image.dart

```diff
@@ -0,0 +1,451 @@
+// Copyright 2017 The Chromium Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+import 'dart:typed_data';
+
+import 'package:flutter/foundation.dart';
+import 'package:flutter/services.dart';
+
+import 'basic.dart';
+import 'framework.dart';
+import 'image.dart';
+import 'ticker_provider.dart';
+
+/// An image that shows a [placeholder] image while the target [image] is
+/// loading, then fades in the new image when it loads.
+///
+/// Use this class to display long-loading images, such as [new NetworkImage],
+/// so that the image appears on screen with a graceful animation rather than a
+/// sudden jerk.
+///
+/// If the [image] emits an [ImageInfo] synchronously, such as when the image
+/// has been loaded and cached, the [image] is displayed immediately and the
+/// [placeholder] is never displayed.
+///
+/// [fadeOutDuration] and [fadeOutCurve] control the fade-out animation of the
+/// placeholder.
+///
+/// [fadeInDuration] and [fadeInCurve] control the fade-in animation of the
+/// target [image].
+///
+/// Prefer a [placeholder] that's already cached so that it is displayed in one
+/// frame. This prevents it from appearing suddenly on the screen.
+///
+/// When [image] changes it is resolved to a new [ImageStream]. If the new
+/// [ImageStream.key] is different this widget subscribes to the new stream and
+/// replaces the displayed image to images emitted by the new stream.
```

Comment:
maybe explicitly mention what happens in the case where placeholder is currently being shown vs image being shown (e.g. if the placeholder is not shown, does the new image gapless-playback itself over the old one, or does it go back to the placeholder then fade in appropriately?)